### ISSUE TITLE
Chromium failing to start with "failed to dup supplied urandom fd: Bad file descriptor"

### ISSUE THREAD

**Issue by jarthod** [rubycdp/ferrum#78]
Hello,

I was seing this error randomly in my builds:
```
Browser did not produce websocket url within 5 seconds
```
The machine was not saturated, of course I tried increasing it (to 20 seconds) but the problem was still the same so it was clear it was not a timing issue.

I tried to enable debug logging but it was way too verbose so impossible to find the error in here. I ended up looking at the code adding just a puts to see the output where the ws URL is missing:
```ruby
        unless ws_url
          @logger.puts(output) if @logger
          puts output
          raise ProcessTimeoutError.new(timeout)
        end
```
And so when the errors occurs I got this:
```
failed to dup supplied urandom fd: Bad file descriptor
```

This error comes from chromium code I guess, but I'm not sure why I have it here and not when I use chromium directly. It doesn't look like Ferrum is "supplying" a custom FD for this.
```
> chromium --version
Chromium 81.0.4044.113 snap
```

I tried starting chromium manually with the exact same parameters:
```
/snap/bin/chromium --headless --disable-gpu --hide-scrollbars --mute-audio --enable-automation --disable-web-security --disable-session-crashed-bubble --disable-breakpad --disable-sync --no-first-run --use-mock-keychain --keep-alive-for-test --disable-popup-blocking --disable-extensions --disable-hang-monitor --disable-features=site-per-process,TranslateUI --disable-translate --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-client-side-phishing-detection --disable-default-apps --disable-dev-shm-usage --disable-ipc-flooding-protection --disable-prompt-on-repost --disable-renderer-backgrounding --force-color-profile=srgb --metrics-recording-only --safebrowsing-disable-auto-update --password-store=basic --window-size=1024,768 --remote-debugging-port=0 --remote-debugging-address=127.0.0.1 --user-data-dir=/tmp/d20200508-7638-17rc0fb
```
Many times but never reproduced this error message.

I run my tests in parallel using `paraspec`, I could reproduce the problem even with a very low concurrency of 2 (on a 6 cores CPU) so the problem is not starting time or slowdown it seems, but maybe concurrency issue somewhere?

I managed to write a small Ferrum script to reproduce the issue using fork:
```ruby
#!/usr/bin/env ruby

require 'ferrum'

10.times do
  fork do
    puts "[#{$$}] Start Ferrum in fork"
    browser = Ferrum::Browser.new
    browser.goto("https://updown.io")
    browser.quit
    puts "[#{$$}] Done."
  end
end
Process.waitall
```
Launching this script reproduce this error every time on my machine (sometimes I also see `/snap/chromium/1117/bin/desktop-launch: line 302: /dev/null: Operation not permitted`)

Any idea about where could this come from and how to fix it?

**Comment by jarthod** [comment]
Ok I investigated this `desktop-launch` error more, I still don't understand why but it seems that when we start more than ~5 chromium process at the same time using the standard Ubuntu snap launcher it generates this errors. I managed to reproduce the issue without Ferrum with just:
```
10.times do
  fork do
    `/snap/bin/chromium --headless`
  end
end
Process.waitall
```
I managed to find the path of the raw chromium binary behind (`/snap/chromium/current/usr/lib/chromium-browser/chrome`)
And with this path it seems to be working fine:

```ruby
#!/usr/bin/env ruby

require 'ferrum'

10.times do
  fork do
    puts "[#{$$}] Start Ferrum in fork"
    browser = Ferrum::Browser.new(browser_path: "/snap/chromium/current/usr/lib/chromium-browser/chrome")
    browser.goto("https://updown.io")
    browser.quit
    puts "[#{$$}] Done."
  end
end
Process.waitall
```
So this again seems to be a problem with Ubuntu snap launcher scripts and not Ferrum.
I'll have to hardcode the binary path then, hopefully this will be useful to other people using Ubuntu.

**Comment by jarthod** [comment]
One thing that might be improved here though is the error reporting, because "Browser did not produce websocket url within 5 seconds" is not very helpful :/ Maybe including at least the first lines of the output in the error message?

**Comment by route** [comment]
Hm maybe. I think `failed to dup supplied urandom fd: Bad file descriptor` is not very helpful either haha, I mean the only thing Ferrum does is spawning new process and it can be a lot of reasons why process hasn't started (mem/security issue or just timings) and I'm not sure Chrome spits something useful or at least something in that case.

**Comment by jarthod** [comment]
Well In this case it's more `/snap/chromium/1117/bin/desktop-launch: line 302: /dev/null: Operation not permitted` that helped me find the issue but what is sure is that "Browser did not produce websocket url within 5 seconds" did not help ;)
It's always a challenge between having clear/simpler error messages and helpful ones, I let you decide what you prefer I just gave my opinion ;)

**Comment by route** [comment]
Got it. Yea I'll try to improve it a bit. Thanks!

**Comment by odlp** [comment]
Hello @route & @jarthod - I've recently experienced a similar issue in that Chrome wouldn't launch and found it difficult to diagnose the root cause. I've opened #92 which proposes including the output captured on the ProcessTimeoutError itself.

**Comment by route** [comment]
I think it can be closed now after merging #92
